File: zwkfb_view/src/main/java/商业/谷歌/安卓/材质/时间选择器/时间模型.java
package 商业.谷歌.安卓.材质.时间选择器;

import static 商业.谷歌.安卓.材质.时间选择器.时间格式.CLOCK_12H;
import static 商业.谷歌.安卓.材质.时间选择器.时间格式.CLOCK_24H;
import static java.util.Calendar.AM;
import static java.util.Calendar.HOUR;
import static java.util.Calendar.PM;

import android.content.res.Resources;
import android.os.Parcel;
import android.os.Parcelable;

import androidx.annotation.IntRange;
import androidx.annotation.Nullable;
import androidx.annotation.StringRes;

import java.util.Arrays;

import com.google.android.material.R;

/** The representation of the TimeModel used by TimePicker views. */
class 时间模型 implements Parcelable {

    public static final String ZERO_LEADING_NUMBER_FORMAT = "%02d";
    public static final String NUMBER_FORMAT = "%d";

    @时间格式
    final int format;

    int hour;
    int minute;

    @时间选择器控件.ActiveSelection
    int selection;
    @时间选择器控件.ClockPeriod
    int period;

    public 时间模型() {
        this(CLOCK_12H);
    }

    public 时间模型(@时间格式 int format) {
        this(0, 0, HOUR, format);
    }

    public 时间模型(int hour, int minute, @时间选择器控件.ActiveSelection int selection, @时间格式 int format) {
        this.hour = hour;
        this.minute = minute;
        this.selection = selection;
        this.format = format;
        period = getPeriod(hour);
    }

    protected 时间模型(Parcel in) {
        this(in.readInt(), in.readInt(), in.readInt(), in.readInt());
    }

    /** Set hour respecting the current clock period */
    public void setHourOfDay(int hour) {
        period = getPeriod(hour);
        this.hour = hour;
    }

    @时间选择器控件.ClockPeriod
    private static int getPeriod(int hourOfDay) {
        return hourOfDay >= 12 ? PM : AM;
    }

    /** Set hour respecting the current clock period */
    public void setHour(int hour) {
        if (format == CLOCK_24H) {
            this.hour = hour;
            return;
        }

        this.hour = hour % 12 + (period == PM ? 12 : 0);
    }

    public void setMinute(@IntRange(from = 0, to = 59) int minute) {
        this.minute = minute % 60;
    }

    public int getHourForDisplay() {
        if (format == CLOCK_24H) {
            return hour % 24;
        }

        if (hour % 12 == 0) {
            return 12;
        }

        if (period == PM) {
            return hour - 12;
        }

        return hour;
    }

    @StringRes
    public int getHourContentDescriptionResId() {
        return format == CLOCK_24H ? R.string.material_hour_24h_suffix : R.string.material_hour_suffix;
    }

    @Override
    public int hashCode() {
        Object[] hashedFields = {format, hour, minute, selection};
        return Arrays.hashCode(hashedFields);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }

        if (!(o instanceof 时间模型)) {
            return false;
        }

        时间模型 that = (时间模型) o;
        return hour == that.hour
                && minute == that.minute
                && format == that.format
                && selection == that.selection;
    }

    @Override
    public int describeContents() {
        return 0;
    }

    @Override
    public void writeToParcel(Parcel dest, int flags) {
        dest.writeInt(hour);
        dest.writeInt(minute);
        dest.writeInt(selection);
        dest.writeInt(format);
    }

    @SuppressWarnings("unused")
    public static final Parcelable.Creator<时间模型> CREATOR =
            new Parcelable.Creator<时间模型>() {
                @Override
                public 时间模型 createFromParcel(Parcel in) {
                    return new 时间模型(in);
                }

                @Override
                public 时间模型[] newArray(int size) {
                    return new 时间模型[size];
                }
            };

    public void setPeriod(@时间选择器控件.ClockPeriod int period) {
        if (period != this.period) {
            this.period = period;
            if (hour < 12 && period == PM) {
                hour += 12;
            } else if (hour >= 12 && period == AM) {
                hour -= 12;
            }
        }
    }

    @Nullable
    public static String formatText(Resources resources, CharSequence text) {
        return formatText(resources, text, ZERO_LEADING_NUMBER_FORMAT);
    }

    @Nullable
    public static String formatText(Resources resources, CharSequence text, String format) {
        try {
            return String.format(resources.getConfiguration().locale, format, Integer.parseInt(String.valueOf(text)));
        } catch (NumberFormatException e) {
            return null;
        }
    }
}

File: zwkfb_view/src/main/java/商业/谷歌/安卓/材质/时间选择器/时间模型.kt
package 商业.谷歌.安卓.材质.时间选择器

import android.content.res.Resources
import android.os.Parcel
import android.os.Parcelable
import androidx.annotation.IntRange
import androidx.annotation.StringRes
import com.google.android.material.R
import java.util.Calendar

/** The representation of the TimeModel used by TimePicker views.  */
internal class 时间模型(
    var hour: Int,
    var minute: Int,
    @field:时间选择器控制器.ActiveSelection @param:时间选择器控制器.ActiveSelection var selection: Int,
    @field:时间格式 @param:时间格式 val format: Int
) : Parcelable {
    val minuteInputValidator: 最大输入验证器
    val hourInputValidator: 最大输入验证器

    @时间选择器控制器.ClockPeriod
    var period: Int = 0

    @JvmOverloads
    constructor(@时间格式 format: Int = 时间格式.CLOCK_12H) : this(0, 0, Calendar.HOUR, format)

    protected constructor(`in`: Parcel) : this(
        `in`.readInt(),
        `in`.readInt(),
        `in`.readInt(),
        `in`.readInt()
    )

    /** Set hour respecting the current clock period  */
    fun setHourOfDay(hour: Int) {
        period = getPeriod(hour)
        this.hour = hour
    }

    /** Set hour respecting the current clock period  */
    @JvmName("setHourInternal")
    fun setHour(hour: Int) {
        if (format == 时间格式.CLOCK_24H) {
            this.hour = hour
            return
        }

        this.hour = hour % 12 + (if (period == Calendar.PM) 12 else 0)
    }

    @JvmName("setMinuteValue")
    fun setMinute(@IntRange(from = 0, to = 59) minute: Int) {
        this.minute = minute % 60
    }

    val hourForDisplay: Int
        get() {
            if (format == 时间格式.CLOCK_24H) {
                return hour % 24
            }

            if (hour % 12 == 0) {
                return 12
            }

            if (period == Calendar.PM) {
                return hour - 12
            }

            return hour
        }

    @get:StringRes
    val hourContentDescriptionResId: Int
        get() = if (format == 时间格式.CLOCK_24H) R.string.material_hour_24h_suffix else R.string.material_hour_suffix

    override fun hashCode(): Int {
        val hashedFields = arrayOf<Any?>(format, hour, minute, selection)
        return hashedFields.contentHashCode()
    }

    override fun equals(o: Any?): Boolean {
        if (this === o) {
            return true
        }

        if (o !is 时间模型) {
            return false
        }

        val that = o
        return hour == that.hour && minute == that.minute && format == that.format && selection == that.selection
    }

    override fun describeContents(): Int {
        return 0
    }

    override fun writeToParcel(dest: Parcel, flags: Int) {
        dest.writeInt(hour)
        dest.writeInt(minute)
        dest.writeInt(selection)
        dest.writeInt(format)
    }

    init {
        period = getPeriod(hour)
        minuteInputValidator = 最大输入验证器(59)
        hourInputValidator = 最大输入验证器(if (format == 时间格式.CLOCK_24H) 23 else 12)
    }

    @JvmName("setPeriodInternal")
    fun setPeriod(@时间选择器控制器.ClockPeriod period: Int) {
        if (period != this.period) {
            this.period = period
            if (hour < 12 && period == Calendar.PM) {
                hour += 12
            } else if (hour >= 12 && period == Calendar.AM) {
                hour -= 12
            }
        }
    }

    companion object {
        const val ZERO_LEADING_NUMBER_FORMAT: String = "%02d"
        const val NUMBER_FORMAT: String = "%d"

        @时间选择器控制器.ClockPeriod
        private fun getPeriod(hourOfDay: Int): Int {
            return if (hourOfDay >= 12) Calendar.PM else Calendar.AM
        }

        @JvmField
        @Suppress("unused")
        val CREATOR: Parcelable.Creator<时间模型?> = object : Parcelable.Creator<时间模型?> {
            override fun createFromParcel(`in`: Parcel): 时间模型 {
                return 时间模型(`in`)
            }

            override fun newArray(size: Int): Array<时间模型?> {
                return arrayOfNulls<时间模型>(size)
            }
        }

        @JvmOverloads
        fun formatText(
            resources: Resources,
            text: CharSequence?,
            format: String = ZERO_LEADING_NUMBER_FORMAT
        ): String? {
            try {
                return String.format(
                    resources.getConfiguration().locale, format, text.toString().toInt()
                )
            } catch (e: NumberFormatException) {
                return null
            }
        }
    }
}
